### PR TITLE
SERVER-98273 Create remediation for measurements are not sorted in increasing order for v2 buckets, or are sorted for v3 buckets

### PR DESCRIPTION
**Pull request by seristof**
Tested with these following test cases:
1. The original repro for SERVER-97441 (had the rebuild the server with a commit before SERVER-97441 was merged):
```
// SERVER-94471 repro (this only works before SERVER-97441 was merged).
// Insert two documents near 1-1-1970
const docs = [
    {"t": ISODate("1969-12-31T23:30:30.001Z")},
    {
        "t": ISODate("1970-01-01T00:15:00.001Z"),
    },
]
assert.commandWorked(tsColl.insert(docs))
// Confirm both measurements are in one bucket.
assert.eq(db["system.buckets.ts"].find({}).toArray().length, 1);
let res = tsColl.validate({full: true});
assert(!res.valid);
runFixBucketVersionMismatchProcedure(tsColl);
res = tsColl.validate({full: true});
assert(res.valid);
tsColl.drop();
```

2. Data with sorted v3 buckets (should be remediated)
```
// Create time-series collection
const collName = 'ts';
const tsColl = db[collName];
const bucketsColl = db.getCollection('system.buckets.' + collName);
const timeField = 't';
tsColl.drop();
assert.commandWorked(db.createCollection(collName, {timeseries: {timeField: timeField}}));

// Get the timeField name of the collection
const tsCollInfo = db.runCommand({listCollections: 1, filter: {name: tsColl.getName()}});
const timeFieldName = tsCollInfo["cursor"]["firstBatch"][0]["options"]["timeseries"]["timeField"];

// Validation failure cases.
dataWithSortedv3Buckets = {
    "_id": ObjectId("65a6eb806ffc9fa4280ecac4"),
    "control": {
        "version": BucketVersion.kCompressedUnsorted,
        "min": {
            "_id": ObjectId("65a6eba7e6d2e848e08c3750"),
            "t": ISODate("2024-01-16T20:48:00Z"),
            "a": 0,
        },
        "max": {
            "_id": ObjectId("65a6eba7e6d2e848e08c3751"),
            "t": ISODate("2024-01-16T20:48:39.448Z"),
            "a": 1,
        },
        "count": NumberInt(2),
    },
    "data": {
        "t": BinData(7, "CQAYhggUjQEAAIAOAAAAAAAAAAA="),
        "a": BinData(7, "AQAAAAAAAAAAAJAuAAAAAAAAAAA="),
        "_id": BinData(7, "BwBlpuun5tLoSOCMN1CALgAAAAAAAAAA"),
    }
};
assert.commandWorked(bucketsColl.insert(dataWithSortedv3Buckets));
res = tsColl.validate({full: true});
assert(!res.valid);
print(tojson(res));
runFixBucketVersionMismatchProcedure(tsColl);
res = tsColl.validate({full: true});
print(tojson(res));
assert(res.valid);
tsColl.drop();
```

3. Data with unsorted v2 buckets (should be remediated)
```
assert.commandWorked(db.createCollection(collName, {timeseries: {timeField: timeField}}));
dataWithUnsortedv2Buckets = {
    "_id": ObjectId("630ea4802093f9983fc394dc"),
    "control": {
        "version": BucketVersion.kCompressedSorted,
        "min": {
            "_id": ObjectId("630fabf7c388456f8aea4f2d"),
            "t": ISODate("2022-08-31T00:00:00.000Z"),
            "a": 0
        },
        "max": {
            "_id": ObjectId("630fabf7c388456f8aea4f2f"),
            "t": ISODate("2022-08-31T00:00:01.000Z"),
            "a": 1
        },
        "count": 2
    },
    "data": {
        "t": BinData(7, "CQDolzLxggEAAID+fAAAAAAAAAA="),
        "_id": BinData(7, "BwBjD6v3w4hFb4rqTy2ATgAAAAAAAAAA"),
        "a": BinData(7, "EAAAAAAAgC4AAAAAAAAAAA==")
    }
};
assert.commandWorked(bucketsColl.insert(dataWithUnsortedv2Buckets))
res = tsColl.validate({full: true});
assert(!res.valid);
runFixBucketVersionMismatchProcedure(tsColl);
res = tsColl.validate({full: true});
assert(res.valid);
tsColl.drop();
```

4. Data with unsorted v3 buckets (should not be remediated, tested to ensure no impacts)
```
// Validation pass cases (ensuring that we don't modify correct buckets).
assert.commandWorked(db.createCollection(collName, {timeseries: {timeField: timeField}}));
dataWithUnsortedv3Buckets = {
    "_id": ObjectId("630ea4802093f9983fc394dc"),
    "control": {
        "version": BucketVersion.kCompressedUnsorted,
        "min": {
            "_id": ObjectId("630fabf7c388456f8aea4f2d"),
            "t": ISODate("2022-08-31T00:00:00.000Z"),
            "a": 0
        },
        "max": {
            "_id": ObjectId("630fabf7c388456f8aea4f2f"),
            "t": ISODate("2022-08-31T00:00:01.000Z"),
            "a": 1
        },
        "count": 2
    },
    "data": {
        "t": BinData(7, "CQDolzLxggEAAID+fAAAAAAAAAA="),
        "_id": BinData(7, "BwBjD6v3w4hFb4rqTy2ATgAAAAAAAAAA"),
        "a": BinData(7, "EAAAAAAAgC4AAAAAAAAAAA==")
    }
};
assert.commandWorked(bucketsColl.insert(dataWithUnsortedv3Buckets))
res = tsColl.validate({full: true});
assert(res.valid);
runFixBucketVersionMismatchProcedure(tsColl);
res = tsColl.validate({full: true});
assert(res.valid);
tsColl.drop();
```

5. Data with sorted v2 buckets (should not be remediated, tested to ensure no impacts)
```
assert.commandWorked(db.createCollection(collName, {timeseries: {timeField: timeField}}));
dataWithSortedv2Buckets = {
    "_id": ObjectId("65a6eb806ffc9fa4280ecac4"),
    "control": {
        "version": BucketVersion.kCompressedSorted,
        "min": {
            "_id": ObjectId("65a6eba7e6d2e848e08c3750"),
            "t": ISODate("2024-01-16T20:48:00Z"),
            "a": 0,
        },
        "max": {
            "_id": ObjectId("65a6eba7e6d2e848e08c3751"),
            "t": ISODate("2024-01-16T20:48:39.448Z"),
            "a": 1,
        },
        "count": NumberInt(2),
    },
    "data": {
        "t": BinData(7, "CQAYhggUjQEAAIAOAAAAAAAAAAA="),
        "a": BinData(7, "AQAAAAAAAAAAAJAuAAAAAAAAAAA="),
        "_id": BinData(7, "BwBlpuun5tLoSOCMN1CALgAAAAAAAAAA"),
    }
};
assert.commandWorked(bucketsColl.insert(dataWithSortedv2Buckets))
res = tsColl.validate({full: true});
assert(res.valid);
runFixBucketVersionMismatchProcedure(tsColl);
res = tsColl.validate({full: true});
assert(res.valid);
tsColl.drop();
```

### CODE REVIEW

File: bucket-version-mismatch/README.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+```
+or 
+```
+..."c":"STORAGE",  "id":6698300, "ctx":"conn9","msg":"Document is not compliant with time-series specifications","attr":{"namespace":"[...]","recordId":[...],"reason":{"code":2,"codeName":"BadValue","errmsg":"Time-series bucket is v3 but has its measurements in-order on time"}}...
+```
+# Remediation
+## Rewrite Bucket Version in a Time Series Collections
+While the script is running, the performance of operations on the time-series collection may be impacted. The script does a scan of the whole collection and performs updates on impacted buckets, which may result in a large load if many buckets are affected.  
+At a high level, the script remediates performance by updating the bucket's version from v2 to v3 and vice versa if a v2 bucket has unsorted data, and if a  from the mixed-schema format to the older schema.  The rewrite is done by unpacking the measurements of the problematic mixed-schema buckets and inserting those measurements back into the collection.
```
Comment (4hp-4int) [inline bucket-version-mismatch/README.md:86]: "The script remediates performance" 

Should this say, "the script remediates validation issues by...."?

Comment (4hp-4int) [reply bucket-version-mismatch/README.md:86]: nit: there is a double space here: "and if a  from" between 'a' and 'from'

File: bucket-version-mismatch/rewrite_timeseries_bucket_version_mismatch.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+function runFixBucketVersionMismatchProcedure(collName) {
+  print(
+      'Checking if the bucket versions match their data in ' + collName +
+      ' ...\n');
+  // Range through all the bucketDocs and change the control version of the
+  // bucket from 2 -> 3 if the data is not sorted or from 3 -> 2 if the data is
+  // sorted.
+  const bucketsColl = db.getCollection('system.buckets.' + collName.getName());
```
Comment (4hp-4int) [inline bucket-version-mismatch/rewrite_timeseries_bucket_version_mismatch.js:70]: When I run this script after changing collName at the top, this is the error I receive:

```
switched to db foo
replset@primary foo> load('rewrite_timeseries_bucket_version_mismatch.js')
Checking if the bucket versions match their data in fix ...

TypeError: collName.getName is not a function
```

I believe line 70 of `bucket-version-mismatch/rewrite_timeseries_bucket_version_mismatch.js` should read: `coll.getName()`  or perhaps just `collName`.

**Review by 4hp-4int — CHANGES_REQUESTED** [review]
I found a few small changes in the README and encountered 1 error when running the remediation script.

**Review by dhly-etc — APPROVED** [review]
One small cleanup on the README, plus other reviewers' finds, but otherwise LGTM. Nice work!

**Review by 4hp-4int — APPROVED** [review]

**Pull request merged**